sqlmigrations: add migration for 19.2-style schema change jobs

thoszhang opened the pull request:
# sqlmigrations: add migration for 19.2-style schema change jobs

This PR adds a migration which upgrades the format of any 19.2 schema
change jobs that remain on a cluster after upgrading to 20.1, so that
those schema changes can run to completion.

For context, note that schema change execution in 19.2 and 20.1 are
mutually incompatible, and 20.1 nodes are forbidden both from initiating
new schema changes before upgrade finalization and from attempting to
run 19.2-style schema changes. Any 19.2 schema changes that linger on
the cluster as a result of being in progress while upgrading the cluster
to have 20.1 binaries must be upgraded to the 20.1 job format before
they can be run in the 20.1 schema change job resumer.

This migration is run inside an async task started in `server.Start()`
after most of the other server state is initialized, and is separate
from the existing startup migrations which must be finished before the
node fully starts up. The migration itself must first wait for the
cluster upgrade to be finalized, to ensure the absence of any 19.2
binaries from the cluster. After upgrade finalization, the migration is
attempted in the background until it finishes successfully. Once any
given job undergoes the migration, it is eligible to be adopted by the
job registry.

The migration has 2 steps: In the first step, we scan all jobs, look up
the table descriptor for each job, and update the job state based on the
schema change-related state on the table descriptor. For jobs that are
waiting for GC for a dropped index, table, etc., we mark the job as
successful and create a GC job. In the second step, we scan all
descriptors and running jobs, and create a new job for each table that
is either being added or has draining names which does not have a
running schema change job. Tables in either of those two states used to
rely on a schema changer in 19.2 to move them out of those intermediate
states, and now require a job in 20.1.

As with the existing startup migrations, we write KVs to the
`system-version/` span to checkpoint the progress of the migration.

Closes #46177.

Release justification: This is needed to support the upgrade path for
the new schema change job.

Release note (general change): Schema changes started in 19.2 will now
be automatically migrated in the background in 20.1 after upgrade
finalization so that they can run to completion.

----

# sql: add schema change migration testing

This commit introduces a testing framework which tests the migration of
schema changes from the 19.2 to the 20.1 format. A test case is created
for each type of schema change so that they can be stressed and
testrace'd individually. For each schema change, the test will:

1. Create a 20.1 schema change.
2. Block the schema change at a certain point in its execution.
3. Mutate the job descriptor and table descriptor such that it appears
as a 19.2 format job. These jobs will not be resumed anymore as 20.1
will refuse to run 19.2 jobs.
4. Verify that the job has been marked as a 19.2 job and is blocked.
5. Run the migration and wait for the migration to complete.
6. Ensure that the schema change completes.

Release justification: testing feature required for release.
Release note: None

----

# testing: increase timeout for CI stress tests

This commit increases the timeout on CI so that make stressrace can
complete. There should be a follow up investigation as to why these
tests take this long to complete.

Release note: None

----

# importccl: set system config trigger in TestGetDescriptorFromDB

The transaction used to write database descriptors in
`TestGetDescriptorFromDB` was missing a call to
`txn.SetSystemConfigTrigger()`. This would be caught by
`CheckEndTxnTrigger` in some cases, in `CreateTestServerParams()`, but
it looks like the error generated by `CheckEndTxnTrigger` to indicate
that the trigger had not been set was consistently being swallowed in
`*txnCommitter.makeTxnCommitExplicitAsync()`, causing the test to pass
anyway. This test started failing with the introduction of the migration
to update schema change jobs, which needs to read all descriptors, and
in doing so somehow causes the error from `CheckEndTxnTrigger` to not be
swallowed anymore. This PR fixes the test.

Release note: None

pbardea committed:
sql: add schema change migration testing

This commit introduces a testing framework which tests the migration of
schema changes from the 19.2 to the 20.1 format. A test case is created
for each type of schema change so that they can be stressed and
testrace'd individually. For each schema change, the test will:

1. Create a 20.1 schema change.
2. Block the schema change at a certain point in its execution.
3. Mutate the job descriptor and table descriptor such that it appears
as a 19.2 format job. These jobs will not be resumed anymore as 20.1
will refuse to run 19.2 jobs.
4. Verify that the job has been marked as a 19.2 job and is blocked.
5. Run the migration and wait for the migration to complete.
6. Ensure that the schema change completes.

Release justification: testing feature required for release.
Release note: None

lucy-zhang committed:
sqlmigrations: add migration for 19.2-style schema change jobs

This PR adds a migration which upgrades the format of any 19.2 schema
change jobs that remain on a cluster after upgrading to 20.1, so that
those schema changes can run to completion.

For context, note that schema change execution in 19.2 and 20.1 are
mutually incompatible, and 20.1 nodes are forbidden both from initiating
new schema changes before upgrade finalization and from attempting to
run 19.2-style schema changes. Any 19.2 schema changes that linger on
the cluster as a result of being in progress while upgrading the cluster
to have 20.1 binaries must be upgraded to the 20.1 job format before
they can be run in the 20.1 schema change job resumer.

This migration is run inside an async task started in `server.Start()`
after most of the other server state is initialized, and is separate
from the existing startup migrations which must be finished before the
node fully starts up. The migration itself must first wait for the
cluster upgrade to be finalized, to ensure the absence of any 19.2
binaries from the cluster. After upgrade finalization, the migration is
attempted in the background until it finishes successfully. Once any
given job undergoes the migration, it is eligible to be adopted by the
job registry.

The migration has 2 steps: In the first step, we scan all jobs, look up
the table descriptor for each job, and update the job state based on the
schema change-related state on the table descriptor. For jobs that are
waiting for GC for a dropped index, table, etc., we mark the job as
successful and create a GC job. In the second step, we scan all
descriptors and running jobs, and create a new job for each table that
is either being added or has draining names which does not have a
running schema change job. Tables in either of those two states used to
rely on a schema changer in 19.2 to move them out of those intermediate
states, and now require a job in 20.1.

As with the existing startup migrations, we write KVs to the
`system-version/` span to checkpoint the progress of the migration.

Release justification: This is needed to support the upgrade path for
the new schema change job.

Release note (general change): Schema changes started in 19.2 will now
be automatically migrated in the background in 20.1 after upgrade
finalization so that they can run to completion.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/46504)
<!-- Reviewable:end -->

lucy-zhang commented:
@pbardea has been working on test on a separate branch which aren't included in the PR yet. We'll amend the PR later today, but I wanted to get this out now.

ajwerner reviewed (changes requested):
Just took a first pass through the migration, still need to read through the testing. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-:-M3SeemnDnH_HYFD1Uu_:b-4vnqz1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @lucy-zhang, @pbardea, and @spaskob)

___
*[pkg/sqlmigrations/migrations.go, line 646 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3SbLhsCmfUdJNzYvAi:-M3SbLhsCmfUdJNzYvAj:b-qfl1rn) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L646)):*
> ```Go
> // is finalized before running the migration. The migration is retried until
> // it succeeds (on any node).
> func (m *Manager) StartSchemaChangeJobMigration(ctx context.Context, ignoreKV bool) error {
> ```

Comment about `ignoreKV` and note that it exists for testing? I have a preference to put it into the TestingKnobs. That way it will totally explicit that this is about testing.
___
*[pkg/sqlmigrations/migrations.go, line 651 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3SaiYk5dq88_3ij1ku:-M3SaiYlECNqjE1AIOyX:b6ujl7) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L651)):*
> ```Go
> 		// First wait for the cluster to finalize the upgrade to 20.1.
> 		waitRetryOpts := retry.Options{
> 			InitialBackoff: 5 * time.Second,
> ```

nit: This whole initial backoff of 5 and then multiply by 2 one time seems complicated for no obvious reason. Why not just InitialBackoff of 10 seconds?
___
*[pkg/sqlmigrations/migrations.go, line 661 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3Sb6FW2l7wfVgR0QpJ:-M3Sb6FW2l7wfVgR0QpK:b11hrc9) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L661)):*
> ```Go
> 			}
> 		}
> 		select {
> ```

Should we do stopper.CancelOnQuiesce for the context so the above loop exits?
___
*[pkg/sqlmigrations/migrations.go, line 685 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3SbdPl3mDX3HcfWhb-:-M3SbdPl3mDX3HcfWhb0:bsgu3jc) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L685)):*
> ```Go
> 			settings:    m.settings,
> 		}
> 		migrationRetryOpts := retry.Options{
> ```

can we refactor the constants here out into a method on the manager to create retry options? I see this is duplicated in a couple of places.
___
*[pkg/sqlmigrations/migrations.go, line 692 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3Sbk_70bXMBOpyJs6a:-M3Sbk_84vcQIzOC8W3n:bostl0s) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L692)):*
> ```Go
> 		startTime := timeutil.Now().String()
> 		for migRetry := retry.Start(migrationRetryOpts); migRetry.Next(); {
> 			if err := migrateSchemaChangeJobs(context.Background(), r, m.jobRegistry); err != nil {
> ```

can you comment about why we're using context.Background here? I might additionally encourage you to create a context with some useful logtags.
___
*[pkg/sqlmigrations/migrations.go, line 713 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3ScELU8xPC8Fw5oQsP:-M3ScELU8xPC8Fw5oQsQ:b-n8cktb) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L713)):*
> ```Go
> // names.
> func schemaChangeMigrationKeyForTable(tableID sqlbase.ID) roachpb.Key {
> 	// TODO (lucy): There is probably a better way to do this.
> ```

why the `/` at all? The encoding is delimiter free as it is. This seems indeed gnarly.  
___
*[pkg/sqlmigrations/migrations.go, line 734 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3SdZ_I8hDYWSbztNqM:-M3SdZ_I8hDYWSbztNqN:bdpuft1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L734)):*
> ```Go
> 	// changes.) This function also handles jobs for indexes waiting for GC,
> 	// stored in GCMutations.
> 	migrateMutationJobForTable := func(txn *kv.Txn, job *jobs.Job, tableDesc *sql.TableDescriptor) error {
> ```

These closures don't really close over much. I could see pulling them out into separate functions. I go back and forth on whether that's better. These things are pretty big so maybe for readability it'd be better.
___
*[pkg/sqlmigrations/migrations.go, line 743 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3Scfno1RSKtiWg3Yi8:-M3Scfnp9PWziilQ1Pxv:bmshtpa) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L743)):*
> ```Go
> 		for i := range tableDesc.MutationJobs {
> 			mutationJob := &tableDesc.MutationJobs[i]
> 			if mutationJob.JobID == *job.ID() {
> ```

total nit: invert the condition and continue.
___
*[pkg/sqlmigrations/migrations.go, line 753 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3Sevg7CDTe01gFZOyn:-M3Sevg7CDTe01gFZOyo:b-et96d1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L753)):*
> ```Go
> 				if err := job.WithTxn(txn).Update(ctx, func(txn *kv.Txn, md jobs.JobMetadata, ju *jobs.JobUpdater) error {
> 					// Update the job details with the table and mutation IDs.
> 					details := md.Payload.UnwrapDetails().(jobspb.SchemaChangeDetails)
> ```

`md.Payload.GetSchemaChange()`?
___
*[pkg/sqlmigrations/migrations.go, line 800 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3Sd3E284HI_nSLw4bx:-M3Sd3E284HI_nSLw4by:bav21km) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L800)):*
> ```Go
> 			gcMutation := &tableDesc.GCMutations[i]
> 			// JobID and dropTime are populated only in 19.2 and earlier versions.
> 			if gcMutation.JobID == *job.ID() {
> ```

same nit about inverting the condition, take it or leave it.
___
*[pkg/sqlmigrations/migrations.go, line 810 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3SfDs64RKMSRvN1I_V:-M3SfDs64RKMSRvN1I_W:b-w9rroc) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L810)):*
> ```Go
> 				log.VEventf(ctx, 2, "job %d: marked as succeeded", *job.ID())
> 
> 				// TODO (lucy): Maybe refactor this so it's created the way that GC jobs
> ```

+1 I find these inline struct creations can break the flow of reading a method because it's not  clear what part of the construction is interesting so I tend to gloss over it. In this case, none of it particularly interesting other than it's creating a gc job.
___
*[pkg/sqlmigrations/migrations.go, line 860 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3Sfzfs6Wied2Q8Mpl_:-M3Sfzfs6Wied2Q8Mpla:bovs0s) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L860)):*
> ```Go
> 	// databases, as well as truncating tables.
> 	migrateDropTablesOrDatabaseJob := func(txn *kv.Txn, job *jobs.Job) error {
> 		details := job.Details().(jobspb.SchemaChangeDetails)
> ```

`job.Payload().GetSchemaChange()`?
___
*[pkg/sqlmigrations/migrations.go, line 916 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3SfwfZEhESndu-8XPq:-M3SfwfZEhESndu-8XPr:b2snwy5) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L916)):*
> ```Go
> 			}
> 			tablesToDrop[i].DropTime = desc.DropTime
> 		}
> ```

same comment about the constructor for this thing.
___
*[pkg/sqlmigrations/migrations.go, line 981 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3SgFhi-1RmdnLOtkZN:-M3SgFhi-1RmdnLOtkZO:bovs0s) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L981)):*
> ```Go
> 				return err
> 			}
> 			details = job.Details().(jobspb.SchemaChangeDetails)
> ```

`job.Payload().GetSchemaChange()`?
___
*[pkg/sqlmigrations/migrations.go, line 1040 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3SgPmw9uCVPsviAv3w:-M3SgPmw9uCVPsviAv3x:bxzanz0) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L1040)):*
> ```Go
> 		// Get all running schema change jobs.
> 		rows, err := r.sqlExecutor.QueryEx(
> 			ctx, "preexisting-jobs", nil, /* txn */
> ```

don't you want to use your `txn` here?
___
*[pkg/sqlmigrations/migrations.go, line 1086 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3Sgs260qfgwH-3jBT2:-M3Sgs260qfgwH-3jBT3:b-5lge1n) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L1086)):*
> ```Go
> 			}
> 
> 			if err := r.db.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
> ```

pulling this logic out into a closure seems positive for readability. 
___
*[pkg/sqlmigrations/migrations.go, line 1111 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3SdwPu196oxJP8I3fe:-M3SdwPu196oxJP8I3ff:b-2rd7vm) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L1111)):*
> ```Go
> 					return nil
> 				}
> 				// TODO (lucy): Is using the root user a good idea?
> ```

maybe the node user? 


<!-- Sent from Reviewable.io -->

lucy-zhang commented:
Status update on the tests: They are approaching their final form. What remains to be done is

- Add tests for the state where a table/index is waiting for GC (trying to test this had uncovered an actual bug in the GC job, fixed in #46691 which is merging soon)
- Fix a flake uncovered by stress testing

thoszhang reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46504)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @pbardea, and @spaskob)

___
*[pkg/sqlmigrations/migrations.go, line 646 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3SbLhsCmfUdJNzYvAi:-M3cpCge5J9xgdZiQrup:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L646)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Comment about `ignoreKV` and note that it exists for testing? I have a preference to put it into the TestingKnobs. That way it will totally explicit that this is about testing.
</blockquote></details>

Done.
___
*[pkg/sqlmigrations/migrations.go, line 651 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3SaiYk5dq88_3ij1ku:-M3cpsW_2UZW5gW9v0-A:bytobau) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L651)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: This whole initial backoff of 5 and then multiply by 2 one time seems complicated for no obvious reason. Why not just InitialBackoff of 10 seconds?
</blockquote></details>

No real reason, I just took this from the retry loop where we attempt to auto-finalize the cluster upgrade. I changed it to have a constant backoff of 10 seconds. I think the only thing that really matters is that the backoff is about the same as in the loop where we retry the upgrade finalization.
___
*[pkg/sqlmigrations/migrations.go, line 661 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3Sb6FW2l7wfVgR0QpJ:-M3csv015MFnggk1_4Q3:b-2bw06o) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L661)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Should we do stopper.CancelOnQuiesce for the context so the above loop exits?
</blockquote></details>

Is there much of an advantage to doing this if we also have `Closer: m.stopper.ShouldQuiesce()` in the retry settings? I thought looking up the cluster version would be pretty quick.
___
*[pkg/sqlmigrations/migrations.go, line 685 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3SbdPl3mDX3HcfWhb-:-M3kv_SW3qO4BPbEkGni:bhkbnfh) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L685)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

can we refactor the constants here out into a method on the manager to create retry options? I see this is duplicated in a couple of places.
</blockquote></details>

Do you mean having a method that takes these values and returns a `retry.Options`? The several usages of the retry options in this file all seem pretty different.
___
*[pkg/sqlmigrations/migrations.go, line 692 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3Sbk_70bXMBOpyJs6a:-M3lVPL-6tOQIfQdYOo2:b-7d58bh) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L692)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

can you comment about why we're using context.Background here? I might additionally encourage you to create a context with some useful logtags.
</blockquote></details>

Having thought about this more, I'm actually not sure what the correct thing is to do here - this originally used `retry.StartWithCtx` with the original `ctx` argument, but then we were seeing context cancellations causing the migration to never retry. Is there a way to always retry the migration until the stopper stops the async task?
___
*[pkg/sqlmigrations/migrations.go, line 713 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3ScELU8xPC8Fw5oQsP:-M3kpnCeAlKsos2EvgCZ:b5iloyn) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L713)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

why the `/` at all? The encoding is delimiter free as it is. This seems indeed gnarly.  
</blockquote></details>

No reason aside from trying to have a prefix ending with `/` to match `system-version/`. I took it out.
___
*[pkg/sqlmigrations/migrations.go, line 734 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3SdZ_I8hDYWSbztNqM:-M3fucRe3bNY7JNxTZGd:b-htkh8m) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L734)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

These closures don't really close over much. I could see pulling them out into separate functions. I go back and forth on whether that's better. These things are pretty big so maybe for readability it'd be better.
</blockquote></details>

I made them into separate functions. You're right that the original function was getting a bit long.
___
*[pkg/sqlmigrations/migrations.go, line 743 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3Scfno1RSKtiWg3Yi8:-M3d1bdC5yD8qbsjE5-b:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L743)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

total nit: invert the condition and continue.
</blockquote></details>

Done.
___
*[pkg/sqlmigrations/migrations.go, line 753 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3Sevg7CDTe01gFZOyn:-M3d1Ncm3Bx12DZ08o9G:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L753)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

`md.Payload.GetSchemaChange()`?
</blockquote></details>

Done.
___
*[pkg/sqlmigrations/migrations.go, line 800 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3Sd3E284HI_nSLw4bx:-M3d1m_z8KTGtkO1hS7M:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L800)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

same nit about inverting the condition, take it or leave it.
</blockquote></details>

Done.
___
*[pkg/sqlmigrations/migrations.go, line 810 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3SfDs64RKMSRvN1I_V:-M3kzQmm53iWBGZP4NxM:b-156afu) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L810)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

+1 I find these inline struct creations can break the flow of reading a method because it's not  clear what part of the construction is interesting so I tend to gloss over it. In this case, none of it particularly interesting other than it's creating a gc job.
</blockquote></details>

Done, I ended up making a new function to do this in `sql`.
___
*[pkg/sqlmigrations/migrations.go, line 860 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3Sfzfs6Wied2Q8Mpl_:-M3fs5SRBRaJ8qqw__UJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L860)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

`job.Payload().GetSchemaChange()`?
</blockquote></details>

Done.
___
*[pkg/sqlmigrations/migrations.go, line 916 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3SfwfZEhESndu-8XPq:-M3lBqQb4EuQIVO9CUTa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L916)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

same comment about the constructor for this thing.
</blockquote></details>

Done.
___
*[pkg/sqlmigrations/migrations.go, line 981 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3SgFhi-1RmdnLOtkZN:-M3fuqg49cqOJF8tgeaG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L981)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

`job.Payload().GetSchemaChange()`?
</blockquote></details>

Done.
___
*[pkg/sqlmigrations/migrations.go, line 1040 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3SgPmw9uCVPsviAv3w:-M3fuwy5234cxTEjrD-t:bqfcuyh) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L1040)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

don't you want to use your `txn` here?
</blockquote></details>

Yeah, thanks.
___
*[pkg/sqlmigrations/migrations.go, line 1086 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3Sgs260qfgwH-3jBT2:-M3ku6rECfV0WgPZFult:b-pz4hkj) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L1086)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

pulling this logic out into a closure seems positive for readability. 
</blockquote></details>

Done. (I moved out everything except reading/writing the record of having created the job for the table.)
___
*[pkg/sqlmigrations/migrations.go, line 1111 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3SdwPu196oxJP8I3fe:-M3krcZj-w_fLYeJqTWd:bs7wumx) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L1111)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

maybe the node user? 
</blockquote></details>

Done. 


<!-- Sent from Reviewable.io -->

lucy-zhang commented:
Update: We're down to fixing test flakes. Here are all the ones we know about:

- On the drop index/table and truncate tests, we usually get a timeout after ~500 runs (which is what's showing up in CI), which is probably related to the GC job somehow.
- Occasionally the tests fail with an error that the job format version is unexpected, so we're probably grabbing the wrong job when validating.
- I saw this failure once:
```
    --- FAIL: TestMigrateDropIndex/drop-index-blocked-at-waiting-for-gc (16.75s)
        schema_change_migrations_test.go:129: job 4: expected status running, got succeeded
```

pbardea committed:
testing: increase timeout for CI stress tests

This commit increases the timeout on CI so that make stressrace can
complete. There should be a follow up investigation as to why these
tests take this long to complete.

Release note: None

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 


Just minor nits and some discussion. Let's do this and clean it up afterwards?

> Reviewed 11 of 15 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-:-M3w_b1N8GtPZvK-05nj:bkbu7n7)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @lucy-zhang, @pbardea, and @spaskob)

___
*[pkg/sql/schema_change_migrations_test.go, line 445 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3wXnWOD3Tv8jHSB2sa:-M3wXnWOD3Tv8jHSB2sb:b-1tp02z) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5430dc2a041bf5bf3969e6063fab449ab869e29/pkg/sql/schema_change_migrations_test.go#L445)):*
> ```Go
> 		// until the new descriptor appears.
> 		var tableDesc *sql.TableDescriptor
> 		for r := retry.Start(base.DefaultRetryOptions()); r.Next(); {
> ```

Is this handled 
___
*[pkg/sql/schema_change_migrations_test.go, line 489 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3wWRnr0tvCC1XNimAR:-M3wWRnr0tvCC1XNimAS:btxt102) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5430dc2a041bf5bf3969e6063fab449ab869e29/pkg/sql/schema_change_migrations_test.go#L489)):*
> ```Go
> 		numJobs = 2
> 	}
> 	migratedCount := 0
> ```

nit: these are zero values. How about:

```
var (
    mu                   sync.Mutex
    migratedCount        int
    doneReverseMigration bool
    ranCancelCommand     bool
    hasCanceled          bool
)
```
___
*[pkg/sql/schema_change_migrations_test.go, line 630 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3wW0-33P9TiD3czFfY:-M3wW0-33P9TiD3czFfZ:bbmxblu) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5430dc2a041bf5bf3969e6063fab449ab869e29/pkg/sql/schema_change_migrations_test.go#L630)):*
> ```Go
> 		description = "adding table 54"
> 	}
> 	if testCase.schemaChangeType != CreateTable {
> ```

nit: `else`? 
___
*[pkg/sql/schema_change_migrations_test.go, line 739 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3wVpOB3-8MYP5aLOv3:-M3wVpOCDmnn0z-v-GVt:b-eibiqq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5430dc2a041bf5bf3969e6063fab449ab869e29/pkg/sql/schema_change_migrations_test.go#L739)):*
> ```Go
> }
> 
> func TestMigrateCreateTable(t *testing.T) {
> ```

nit: the below tests all feel incredibly repetitive. There should be an easy way to clean them up as subtests. Not blocking the review on that though.
___
*[pkg/sqlmigrations/migrations.go, line 661 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3Sb6FW2l7wfVgR0QpJ:-M3wYtL06bvMB_Tb2qTV:b-fyh5d7) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L661)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Is there much of an advantage to doing this if we also have `Closer: m.stopper.ShouldQuiesce()` in the retry settings? I thought looking up the cluster version would be pretty quick.
</blockquote></details>

Missed the `Closer`
___
*[pkg/sqlmigrations/migrations.go, line 685 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3SbdPl3mDX3HcfWhb-:-M3wYxuR0BgmkTcvwH8U:b6zl6ex) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L685)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Do you mean having a method that takes these values and returns a `retry.Options`? The several usages of the retry options in this file all seem pretty different.
</blockquote></details>

I was. Fair enough. I saw one other usage that seemed similar. 
___
*[pkg/sqlmigrations/migrations.go, line 692 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3Sbk_70bXMBOpyJs6a:-M3wZEgSEFwdg8Iy1wo6:b-cna0go) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L692)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Having thought about this more, I'm actually not sure what the correct thing is to do here - this originally used `retry.StartWithCtx` with the original `ctx` argument, but then we were seeing context cancellations causing the migration to never retry. Is there a way to always retry the migration until the stopper stops the async task?
</blockquote></details>

Using effectively context.Background with a log tag seems right. You can connect the context to the stopper with `WithCloseOnQuiesce`. 
___
*[pkg/sqlmigrations/migrations.go, line 713 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3ScELU8xPC8Fw5oQsP:-M3wZ_km56WCS6JcQsHi:bw76d4b) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L713)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

No reason aside from trying to have a prefix ending with `/` to match `system-version/`. I took it out.
</blockquote></details>

Our pretty printing of roachpb.Key should print the delimiters as `/`
___
*[pkg/sqlmigrations/migrations.go, line 1113 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3w_Gtq7rE6bO5Cgy58:-M3w_Gtq7rE6bO5Cgy59:b16dx4y) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5430dc2a041bf5bf3969e6063fab449ab869e29/pkg/sqlmigrations/migrations.go#L1113)):*
> ```Go
> 	// of the tables have already been GC'ed at this point? In 19.2, each table
> 	// advances separately through the stages of being dropped, so it should be
> 	// possible for some tables to still be draining names while others have
> ```

What's the worst case here? 


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 


Just minor nits and some discussion. Let's do this and clean it up afterwards?

> Reviewed 11 of 15 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-:-M3w_b1N8GtPZvK-05nj:bkbu7n7)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @lucy-zhang, @pbardea, and @spaskob)

___
*[pkg/sql/schema_change_migrations_test.go, line 445 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3wXnWOD3Tv8jHSB2sa:-M3wXnWOD3Tv8jHSB2sb:b-1tp02z) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5430dc2a041bf5bf3969e6063fab449ab869e29/pkg/sql/schema_change_migrations_test.go#L445)):*
> ```Go
> 		// until the new descriptor appears.
> 		var tableDesc *sql.TableDescriptor
> 		for r := retry.Start(base.DefaultRetryOptions()); r.Next(); {
> ```

Is this handled 
___
*[pkg/sql/schema_change_migrations_test.go, line 489 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3wWRnr0tvCC1XNimAR:-M3wWRnr0tvCC1XNimAS:btxt102) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5430dc2a041bf5bf3969e6063fab449ab869e29/pkg/sql/schema_change_migrations_test.go#L489)):*
> ```Go
> 		numJobs = 2
> 	}
> 	migratedCount := 0
> ```

nit: these are zero values. How about:

```
var (
    mu                   sync.Mutex
    migratedCount        int
    doneReverseMigration bool
    ranCancelCommand     bool
    hasCanceled          bool
)
```
___
*[pkg/sql/schema_change_migrations_test.go, line 630 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3wW0-33P9TiD3czFfY:-M3wW0-33P9TiD3czFfZ:bbmxblu) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5430dc2a041bf5bf3969e6063fab449ab869e29/pkg/sql/schema_change_migrations_test.go#L630)):*
> ```Go
> 		description = "adding table 54"
> 	}
> 	if testCase.schemaChangeType != CreateTable {
> ```

nit: `else`? 
___
*[pkg/sql/schema_change_migrations_test.go, line 739 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3wVpOB3-8MYP5aLOv3:-M3wVpOCDmnn0z-v-GVt:b-eibiqq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5430dc2a041bf5bf3969e6063fab449ab869e29/pkg/sql/schema_change_migrations_test.go#L739)):*
> ```Go
> }
> 
> func TestMigrateCreateTable(t *testing.T) {
> ```

nit: the below tests all feel incredibly repetitive. There should be an easy way to clean them up as subtests. Not blocking the review on that though.
___
*[pkg/sqlmigrations/migrations.go, line 661 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3Sb6FW2l7wfVgR0QpJ:-M3wYtL06bvMB_Tb2qTV:b-fyh5d7) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L661)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Is there much of an advantage to doing this if we also have `Closer: m.stopper.ShouldQuiesce()` in the retry settings? I thought looking up the cluster version would be pretty quick.
</blockquote></details>

Missed the `Closer`
___
*[pkg/sqlmigrations/migrations.go, line 685 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3SbdPl3mDX3HcfWhb-:-M3wYxuR0BgmkTcvwH8U:b6zl6ex) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L685)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Do you mean having a method that takes these values and returns a `retry.Options`? The several usages of the retry options in this file all seem pretty different.
</blockquote></details>

I was. Fair enough. I saw one other usage that seemed similar. 
___
*[pkg/sqlmigrations/migrations.go, line 692 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3Sbk_70bXMBOpyJs6a:-M3wZEgSEFwdg8Iy1wo6:b-cna0go) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L692)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Having thought about this more, I'm actually not sure what the correct thing is to do here - this originally used `retry.StartWithCtx` with the original `ctx` argument, but then we were seeing context cancellations causing the migration to never retry. Is there a way to always retry the migration until the stopper stops the async task?
</blockquote></details>

Using effectively context.Background with a log tag seems right. You can connect the context to the stopper with `WithCloseOnQuiesce`. 
___
*[pkg/sqlmigrations/migrations.go, line 713 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3ScELU8xPC8Fw5oQsP:-M3wZ_km56WCS6JcQsHi:bw76d4b) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L713)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

No reason aside from trying to have a prefix ending with `/` to match `system-version/`. I took it out.
</blockquote></details>

Our pretty printing of roachpb.Key should print the delimiters as `/`
___
*[pkg/sqlmigrations/migrations.go, line 1113 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3w_Gtq7rE6bO5Cgy58:-M3w_Gtq7rE6bO5Cgy59:b16dx4y) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5430dc2a041bf5bf3969e6063fab449ab869e29/pkg/sqlmigrations/migrations.go#L1113)):*
> ```Go
> 	// of the tables have already been GC'ed at this point? In 19.2, each table
> 	// advances separately through the stages of being dropped, so it should be
> 	// possible for some tables to still be draining names while others have
> ```

What's the worst case here? 


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 


Just minor nits and some discussion. Let's do this and clean it up afterwards?

> Reviewed 11 of 15 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-:-M3w_b1N8GtPZvK-05nj:bkbu7n7)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @lucy-zhang, @pbardea, and @spaskob)

___
*[pkg/sql/schema_change_migrations_test.go, line 445 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3wXnWOD3Tv8jHSB2sa:-M3wXnWOD3Tv8jHSB2sb:b-1tp02z) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5430dc2a041bf5bf3969e6063fab449ab869e29/pkg/sql/schema_change_migrations_test.go#L445)):*
> ```Go
> 		// until the new descriptor appears.
> 		var tableDesc *sql.TableDescriptor
> 		for r := retry.Start(base.DefaultRetryOptions()); r.Next(); {
> ```

Is this handled 
___
*[pkg/sql/schema_change_migrations_test.go, line 489 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3wWRnr0tvCC1XNimAR:-M3wWRnr0tvCC1XNimAS:btxt102) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5430dc2a041bf5bf3969e6063fab449ab869e29/pkg/sql/schema_change_migrations_test.go#L489)):*
> ```Go
> 		numJobs = 2
> 	}
> 	migratedCount := 0
> ```

nit: these are zero values. How about:

```
var (
    mu                   sync.Mutex
    migratedCount        int
    doneReverseMigration bool
    ranCancelCommand     bool
    hasCanceled          bool
)
```
___
*[pkg/sql/schema_change_migrations_test.go, line 630 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3wW0-33P9TiD3czFfY:-M3wW0-33P9TiD3czFfZ:bbmxblu) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5430dc2a041bf5bf3969e6063fab449ab869e29/pkg/sql/schema_change_migrations_test.go#L630)):*
> ```Go
> 		description = "adding table 54"
> 	}
> 	if testCase.schemaChangeType != CreateTable {
> ```

nit: `else`? 
___
*[pkg/sql/schema_change_migrations_test.go, line 739 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3wVpOB3-8MYP5aLOv3:-M3wVpOCDmnn0z-v-GVt:b-eibiqq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5430dc2a041bf5bf3969e6063fab449ab869e29/pkg/sql/schema_change_migrations_test.go#L739)):*
> ```Go
> }
> 
> func TestMigrateCreateTable(t *testing.T) {
> ```

nit: the below tests all feel incredibly repetitive. There should be an easy way to clean them up as subtests. Not blocking the review on that though.
___
*[pkg/sqlmigrations/migrations.go, line 661 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3Sb6FW2l7wfVgR0QpJ:-M3wYtL06bvMB_Tb2qTV:b-fyh5d7) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L661)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Is there much of an advantage to doing this if we also have `Closer: m.stopper.ShouldQuiesce()` in the retry settings? I thought looking up the cluster version would be pretty quick.
</blockquote></details>

Missed the `Closer`
___
*[pkg/sqlmigrations/migrations.go, line 685 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3SbdPl3mDX3HcfWhb-:-M3wYxuR0BgmkTcvwH8U:b6zl6ex) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L685)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Do you mean having a method that takes these values and returns a `retry.Options`? The several usages of the retry options in this file all seem pretty different.
</blockquote></details>

I was. Fair enough. I saw one other usage that seemed similar. 
___
*[pkg/sqlmigrations/migrations.go, line 692 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3Sbk_70bXMBOpyJs6a:-M3wZEgSEFwdg8Iy1wo6:b-cna0go) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L692)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Having thought about this more, I'm actually not sure what the correct thing is to do here - this originally used `retry.StartWithCtx` with the original `ctx` argument, but then we were seeing context cancellations causing the migration to never retry. Is there a way to always retry the migration until the stopper stops the async task?
</blockquote></details>

Using effectively context.Background with a log tag seems right. You can connect the context to the stopper with `WithCloseOnQuiesce`. 
___
*[pkg/sqlmigrations/migrations.go, line 713 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3ScELU8xPC8Fw5oQsP:-M3wZ_km56WCS6JcQsHi:bw76d4b) ([raw file](https://github.com/cockroachdb/cockroach/blob/c75e6f788903d3026ee46def7b7ef5d882433cd7/pkg/sqlmigrations/migrations.go#L713)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

No reason aside from trying to have a prefix ending with `/` to match `system-version/`. I took it out.
</blockquote></details>

Our pretty printing of roachpb.Key should print the delimiters as `/`
___
*[pkg/sqlmigrations/migrations.go, line 1113 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/46504#-M3w_Gtq7rE6bO5Cgy58:-M3w_Gtq7rE6bO5Cgy59:b16dx4y) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5430dc2a041bf5bf3969e6063fab449ab869e29/pkg/sqlmigrations/migrations.go#L1113)):*
> ```Go
> 	// of the tables have already been GC'ed at this point? In 19.2, each table
> 	// advances separately through the stages of being dropped, so it should be
> 	// possible for some tables to still be draining names while others have
> ```

What's the worst case here? 


<!-- Sent from Reviewable.io -->

lucy-zhang commented:
Reviewable is not letting me reply inline (probably due to the earlier GitHub downtime), but I addressed all the comments except that
- I haven't attempted to unify all the tests
- I realized we currently throw an error unconditionally when we can't find a table descriptor for a table to be GC'ed. I think this can be easily fixed (and manually tested; all we need to do is make a database with multiple tables and set just one table to be GCed immediately). Writing a DROP DATABASE test also shouldn't be too bad, but I don't want to do it in this PR.

Also, we have a few more new failures in CI which seem pretty tractable. I'll get back to looking at all this tomorrow.

pbardea commented:
As an update, the tests were unified and the CI failures were addressed.

lucy-zhang committed:
importccl: set system config trigger in TestGetDescriptorFromDB

The transaction used to write database descriptors in
`TestGetDescriptorFromDB` was missing a call to
`txn.SetSystemConfigTrigger()`. This would be caught by
`CheckEndTxnTrigger` in some cases, in `CreateTestServerParams()`, but
it looks like the error generated by `CheckEndTxnTrigger` to indicate
that the trigger had not been set was consistently being swallowed in
`*txnCommitter.makeTxnCommitExplicitAsync()`, causing the test to pass
anyway. This test started failing with the introduction of the migration
to update schema change jobs, which needs to read all descriptors, and
in doing so somehow causes the error from `CheckEndTxnTrigger` to not be
swallowed anymore. This PR fixes the test.

Release note: None

lucy-zhang commented:
I added another commit to fix `TestGetDescriptorFromDB`. Details are in the updated PR description.

lucy-zhang commented:
bors r+

lucy-zhang commented:
`TestZip` failed, I think because we wrote more heap profiles than usual.
```
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -47,3 +47,8 @@
        	            	 requesting heap profile for node 1... writing: debug/nodes/1/heap.pprof
        	            	-requesting heap files for node 1... 0 found
        	            	+requesting heap files for node 1... 5 found
        	            	+writing: debug/nodes/1/heapprof/memprof.000000000027009408_2020-04-04T01_42_11.198.pprof
        	            	+writing: debug/nodes/1/heapprof/memprof.000000000032986616_2020-04-04T01_42_31.198.pprof
        	            	+writing: debug/nodes/1/heapprof/memprof.000000000042765320_2020-04-04T01_42_51.199.pprof
        	            	+writing: debug/nodes/1/heapprof/memprof.000000000045172096_2020-04-04T01_43_11.199.pprof
        	            	+writing: debug/nodes/1/heapprof/memprof.000000000045986520_2020-04-04T01_43_31.199.pprof
        	            	 requesting goroutine files for node 1... 0 found
```
I stressed this for ~1000 runs on my laptop and couldn't reproduce, so I'm going to attribute this to a fluke rather than a consistent problem with excess memory usage.

lucy-zhang commented:
bors r+
